File: 0141-linked-list-cycle/0141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if not head: return False
        a, b = head, head.next

        while a != b:
            a = a.next
            
            if not b:
                return False
            b = b.next
            if not b:
                return False

            b = b.next

        return True

File: 0141-linked-list-cycle/0141-linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (head == nullptr) return false;
        ListNode* slow = head;
        ListNode* fast = head->next;
    
        while (fast != nullptr) {
            if (slow == fast) {
                return true;
            }
            fast = fast->next;
            if (fast == nullptr) {
                break;
            }
            fast = fast->next;
            slow = slow->next;
        }
        return false;
    }
};